File: server/src/api/reading/repo/user-book.repository.ts
import type { UserBook, UserBookUpdate } from "@/api/reading/model/user-book.model";
import { Query, SortFields } from '@/core/api/commonModel'

import { db } from '@/drizzle'
import { book, userBook, note } from '@/drizzle/schema'
import { and, eq, getTableColumns } from 'drizzle-orm'
import { buildFilters, buildOrders, parseColumns } from '@/core/utils/drizzle'
import { randomUUID } from 'node:crypto'

export class UserBookRepository {

  async create(data: UserBook) {
    data.id = data.id || randomUUID();
    return db.insert(userBook).values(data).returning().get();
  }

  async findById(id: string) {
    return db.select().from(userBook).where(eq(userBook.id, id)).get();
  }

  async findDetailsById(id: string) {
    return db.select(this.getDetailsColumns())
      .from(userBook)
      .leftJoin(book, eq(userBook.bookId, book.id))
      .where(eq(userBook.id, id))
      .get();
  }

  async findDetailsByBook(userId: string, bookId: string) {
    return db.select(this.getDetailsColumns())
      .from(userBook)
      .leftJoin(book, eq(userBook.bookId, book.id))
      .where(
        and(
          eq(userBook.userId, userId),
          eq(userBook.bookId, bookId)
        )
      )
      .get();
  }

  async find(where: any) {
    return db.select().from(userBook).all();
  }

  async update(data: UserBookUpdate) {
    const id = data.id || '';
    await db.update(userBook).set(data).where(eq(userBook.id, id));
    return this.findById(id)
  }

  async delete(id: string) {
    return db.delete(userBook).where(eq(userBook.id, id));
  }

  async getAll(userId: string) {
    return db.select().from(note).where(eq(note.userId, userId)).all();
  }

  async queryAll(query: Record<string, string>) {
    const filters = [];
    if (query?.bookId) {
      filters.push(eq(userBook.bookId, query.bookId));
    }

    return db.select().from(userBook)
      .where(and(...filters))
      .all();
  }

  async query(query: Query) {
    const filters = buildFilters(userBook, ['bookId', 'userId'], query.condition);
    const orders = buildOrders(userBook, [...SortFields], query.sort);

    return db.select().from(userBook)
      .where(and(...filters))
      .limit(query.pageSize)
      .offset((query.pageIndex - 1) * query.pageSize)
      .orderBy(...orders);
  }

  async queryDetails(query: Query) {
    const userBookFilters = buildFilters(userBook, ['userId'], query.condition);
    const bookFilters = buildFilters(book, ['title'], query.condition);
    const filters = [...userBookFilters, ...bookFilters];

    const userBookOrders = buildOrders(userBook, [...SortFields], query.sort);
    const bookOrders = buildOrders(book, ['title'], query.sort);
    const orders = [...userBookOrders, ...bookOrders];

    return db.select(this.getDetailsColumns()).from(userBook)
      .leftJoin(book, eq(userBook.bookId, book.id))
      .where(and(...filters))
      .limit(query.pageSize)
      .offset((query.pageIndex - 1) * query.pageSize)
      .orderBy(...orders);
  }

  getDetailsColumns() {
    return {
      ...getTableColumns(userBook),
      owner: book.userId,
      title: book.title,
      path: book.path,
      fileName: book.fileName,
      coverName: book.coverName,
      author: book.author,
      language: book.language,
      description: book.description,
      publisher: book.publisher,
      published: book.published,
      scope: book.scope,
      bookRating: book.rating,
    };
  }
}

File: backend/app/api/repos/user_book_repository.py
from sqlalchemy import func, text
from sqlmodel import select
from uuid import UUID

from app.api.models.book import Book
from app.api.models.query import PaginationQuery, QueryResult
from app.api.models.user_book import UserBook
from app.api.repos.base_repository import BaseRepository
from app.libs.db_helper import DbHelper


class UserBookRepository(BaseRepository[UserBook]):
    def __init__(self, model, session):
        super().__init__(model, session)

    def get_extension_stats(self, workspace_id: UUID):
        stmt = (
            select(Book.extension.label("type"), func.count().label("count"))
            .select_from(UserBook)
            .join(Book, Book.id == UserBook.book_id, isouter=True)
            .where(UserBook.workspace_id == str(workspace_id))
            .group_by(Book.extension)
        )
        # print(stmt.compile(compile_kwargs={"literal_binds": True}))
        result = self.session.exec(stmt)
        return [{"type": row.type, "count": row.count} for row in result.all()]

    def get_status_stats(self, workspace_id: UUID):
        stmt = (
            select(UserBook.reading_status.label("status"), func.count().label("count"))
            .where(UserBook.workspace_id == str(workspace_id))
            .group_by(UserBook.reading_status)
        )
        print(stmt.compile(compile_kwargs={"literal_binds": True}))
        result = self.session.exec(stmt)
        return [{"status": row.status, "count": row.count} for row in result.all()]

    def get_details(self, id: UUID) -> dict | None:
        stmt = (
            select(UserBook, Book)
            .join(Book, Book.id == UserBook.book_id)
            .where(UserBook.id == id)
        )
        result = self.session.exec(stmt).first()
        if result:
            user_book, book = result
            return self.build_details(user_book, book)
        return None

    def query_details(self, query: PaginationQuery) -> QueryResult:
        # 1. Filters
        filter_mapping = {
            UserBook: ['workspace_id', 'user_id', 'reading_status'],
            Book: ['title', 'extension'],
        }
        filters = DbHelper.build_filters(filter_mapping, query.condition)

        # 2. stmt
        stmt = (select(UserBook, Book)
            .join(Book, Book.id == UserBook.book_id)
        )
        count_stmt = (select(func.count())
            .select_from(UserBook)
            .join(Book, Book.id == UserBook.book_id)
        )
        if filters:
            stmt = stmt.where(*filters)
            count_stmt = count_stmt.where(*filters)

        # 3. Sort
        stmt = DbHelper.apply_sort(stmt, [UserBook, Book], query.sort)

        # 4. Pagination
        stmt = DbHelper.apply_pagination(stmt, query.pageIndex, query.pageSize)

        # 5. Query
        total = self.session.exec(count_stmt).one()
        rows = [
            self.build_details(user_book, book)
            for user_book, book in self.session.exec(stmt).all()
        ]
        return QueryResult(
            total=total,
            list=rows,
            pageSize=query.pageSize,
            pageIndex=query.pageIndex,
        )

    @staticmethod
    def build_details(workspace_book: UserBook, book: Book) -> dict:
        return {
            **workspace_book.model_dump(),
            "owner": book.user_id,
            "title": book.title,
            "path": book.path,
            "file_name": book.file_name,
            "cover_name": book.cover_name,
            "author": book.author,
            "language": book.language,
            "description": book.description,
            "extension": book.extension,
            "publisher": book.publisher,
            "published": book.published,
            "scope": book.scope,
            "book_rating": book.rating,
        }
